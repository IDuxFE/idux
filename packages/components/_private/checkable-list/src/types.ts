/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
