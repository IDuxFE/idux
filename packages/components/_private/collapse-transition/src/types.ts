/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
