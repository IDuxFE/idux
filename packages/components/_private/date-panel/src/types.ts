/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
