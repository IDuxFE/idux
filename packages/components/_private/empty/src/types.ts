/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
