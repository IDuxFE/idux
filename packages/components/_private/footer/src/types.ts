/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
