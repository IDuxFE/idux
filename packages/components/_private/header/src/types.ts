/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { HeaderProps as IxHeaderProps } from '@idux/components/header'
import type { VNode } from 'vue'

export interface HeaderProps {
  closable?: boolean
  closeIcon?: string | VNode
  header?: string | IxHeaderProps
  size?: 'lg' | 'md' | 'sm'
  onClose?: (evt: Event) => void
}
