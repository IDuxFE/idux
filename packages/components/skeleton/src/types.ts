/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

type SkeletonType = 'text' | 'rect' | 'round' | 'circle'

export const skeletonProps = {
  animated: IxPropTypes.bool,
  loading: IxPropTypes.bool.def(true),
  type: IxPropTypes.oneOf<SkeletonType>(['round', 'circle', 'text', 'rect']).def('text'),
  width: IxPropTypes.oneOfType([String, Number]),
  height: IxPropTypes.oneOfType([String, Number]),
  repeat: IxPropTypes.number.def(1),
}

export type SkeletonProps = IxInnerPropTypes<typeof skeletonProps>
export type SkeletonPublicProps = IxPublicPropTypes<typeof skeletonProps>
export type SkeletonComponent = DefineComponent<Omit<HTMLAttributes, keyof SkeletonPublicProps> & SkeletonPublicProps>
export type SkeletonInstance = InstanceType<DefineComponent<SkeletonProps>>
