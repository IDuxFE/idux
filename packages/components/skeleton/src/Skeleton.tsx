/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass } from 'vue'

import { convertCssPixel } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { useThemeToken } from '@idux/components/theme'

import { skeletonProps } from './types'
import { getThemeTokens } from '../theme'

export default defineComponent({
  name: 'IxSkeleton',
  inheritAttrs: false,
  props: skeletonProps,
  setup(props, { slots, attrs }) {
    const common = useGlobalConfig('common')
    const { globalHashId, hashId, registerToken } = useThemeToken('skeleton')
    registerToken(getThemeTokens)

    const mergedPrefixCls = computed(() => `${common.prefixCls}-skeleton`)
    const config = useGlobalConfig('skeleton')
    const loaderClass = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${prefixCls}-loader`]: true,
        [`${prefixCls}-${props.type}`]: true,
        [`${prefixCls}-animated`]: props.animated ?? config.animated,
      })
    })
    const loaderStyle = computed(() => {
      const { width, height, type } = props

      if (type === 'circle') {
        const size = convertCssPixel(width || height)
        return {
          width: size,
          height: size,
        }
      }
      return {
        width: convertCssPixel(width),
        height: convertCssPixel(height),
      }
    })

    return () => {
      if (!props.loading) {
        return slots.default?.()
      }

      const prefixCls = mergedPrefixCls.value
      const loader = <div class={loaderClass.value} style={loaderStyle.value}></div>
      const loaderList = Array.from({ length: props.repeat }).map(() => loader)

      return (
        <div class={[prefixCls, globalHashId.value, hashId.value]} {...attrs}>
          {loaderList}
        </div>
      )
    }
  },
})
