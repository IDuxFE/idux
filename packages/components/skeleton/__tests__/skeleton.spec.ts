import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Skeleton from '../src/Skeleton'
import { SkeletonProps } from '../src/types'

describe('Skeleton', () => {
  const SkeletonMount = (options?: MountingOptions<Partial<SkeletonProps>>) =>
    mount(Skeleton, { ...(options as MountingOptions<SkeletonProps>) })

  const sizeWork = (sizeName: 'width' | 'height') => {
    test(`props ${sizeName} work`, async () => {
      const wrapper = SkeletonMount({
        props: {
          [sizeName]: 40,
        },
      })
      const getLoaderDom = () => wrapper.find('.ix-skeleton-loader').element as HTMLElement

      let loaderDom = getLoaderDom()
      expect(loaderDom.style[sizeName]).toBe('40px')

      await wrapper.setProps({ [sizeName]: '50px' })
      loaderDom = getLoaderDom()
      expect(loaderDom.style[sizeName]).toBe('50px')

      await wrapper.setProps({ [sizeName]: '30%' })
      loaderDom = getLoaderDom()
      expect(loaderDom.style[sizeName]).toBe('30%')
    })
  }

  renderWork<SkeletonProps>(Skeleton)
  sizeWork('width')
  sizeWork('height')

  test('props animated work', async () => {
    const wrapper = SkeletonMount()

    expect(wrapper.find('.ix-skeleton-loader').classes()).toContain('ix-skeleton-animated')

    await wrapper.setProps({ animated: false })
    expect(wrapper.find('.ix-skeleton-loader').classes()).not.toContain('ix-skeleton-animated')
  })
  test('props type work', async () => {
    const wrapper = SkeletonMount()

    expect(wrapper.find('.ix-skeleton-loader').classes()).toContain('ix-skeleton-text')

    await wrapper.setProps({ type: 'rect' })
    expect(wrapper.find('.ix-skeleton-loader').classes()).toContain('ix-skeleton-rect')

    await wrapper.setProps({ type: 'round' })
    expect(wrapper.find('.ix-skeleton-loader').classes()).toContain('ix-skeleton-round')

    await wrapper.setProps({ type: 'circle' })
    expect(wrapper.find('.ix-skeleton-loader').classes()).toContain('ix-skeleton-circle')
  })
  test('props loading work', async () => {
    const wrapper = SkeletonMount({
      slots: {
        default: () => h('div', 'skeletonSlots'),
      },
    })

    expect(wrapper.find('.ix-skeleton-loader').exists()).toBeTruthy()
    expect(wrapper.text()).not.toBe('skeletonSlots')

    await wrapper.setProps({ loading: false })
    expect(wrapper.find('.ix-skeleton-loader').exists()).toBeFalsy()
    expect(wrapper.text()).toBe('skeletonSlots')
  })
  test('props repeat work', () => {
    const wrapper = SkeletonMount({
      props: {
        repeat: 3,
      },
    })

    expect(wrapper.findAll('.ix-skeleton-loader').length).toBe(3)
  })
})
