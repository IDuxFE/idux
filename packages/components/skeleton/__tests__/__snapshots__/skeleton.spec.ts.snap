// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton render work 1`] = `
"<div class=\\"ix-skeleton\\">
  <div class=\\"ix-skeleton-loader ix-skeleton-text ix-skeleton-animated\\"></div>
</div>"
`;
